R/utils-callbacks.R

```diff
@@ -0,0 +1,126 @@
+
+      # Can't have more than one required arg and required arg must be first
+      is_required <- map2_lgl(
+        arg_names,
+        fn_fmls(callback),
+        function(arg, val) {
+          arg != "..." && is.symbol(val) && identical(val, quote(expr = ))
+        }
+      )
+
+      ok_args <-
+        sum(is_required) == 0 ||
+        (sum(is_required) == 1 && is_required[1])
+
+      if (!ok_args) {
+        cli::cli_abort(
+          "Only the first argument of {.var callback} can be required."
+        )
+      }
+
+      id <- private$next_id()
+      private$callbacks[[as.character(id)]] <- callback
+
+      fn_remove <- function() {
+        private$callbacks[[as.character(id)]] <- NULL
+        invisible(NULL)
+      }
+      invisible(fn_remove)
+    },
+
+    #' @description
+    #' Invoke all registered callbacks with the provided arguments. Callbacks
+    #' are invoked in reverse order of registration (last-in first-evaluated).
+    #'
+    #' @param data An input, such as a list or object, to pass to the callbacks.
+    #' @returns Nothing, callbacks are invoked for side effects).
+    invoke = function(data) {
+      if (length(private$callbacks) == 0) {
+        return(invisible(NULL))
+      }
+
+      # Invoke callbacks in reverse insertion order
+      for (id in rev(as.integer(names(private$callbacks)))) {
+        res <- exec(private$callbacks[[as.character(id)]], data)
+        if (promises::is.promise(res)) {
+          cli::cli_abort(c(
+            "Can't use async callbacks with `$chat()` or `$stream()`.",
+            i = "Async callbacks are supported, but you must use `$chat_async()` or `$stream_async()`."
+          ))
+        }
+      }
+
+      invisible(NULL)
+    },
+
+    #' @description
+    #' Invoke all registered callbacks asynchronously with the provided
+    #' arguments. As with `$invoke()`, callbacks are invoked in reverse order of
+    #' registration (last-in first-evaluated).
+    #'
+    #' @param data An input, such as a list or object, to pass to the callbacks.
+    #' @returns Nothing, callbacks are invoked for side effects).
+    invoke_async = async_method(function(self, private, data) {
+      if (length(private$callbacks) == 0) {
+        return(invisible(NULL))
+      }
+
+      # Invoke callbacks in reverse insertion order
+      for (id in rev(as.integer(names(private$callbacks)))) {
+        coro::await(exec(private$callbacks[[as.character(id)]], data))
+      }
+
+      invisible(NULL)
+    }),
+
+    #' @description Get the number of registered callbacks.
+    #' @return Integer count of callbacks.
+    count = function() {
+      length(private$callbacks)
+    },
+
+    #' @description Clear all registered callbacks.
+    clear = function() {
+      private$callbacks <- list()
+      invisible(NULL)
+    }
+  ),
+
+  private = list(
+    callbacks = list(),
+    id = 1L,
+
+    next_id = function() {
+      id <- private$id
+      private$id <- private$id + 1L
+      id
```

Comment:
```suggestion
      as.character(id)
```

Simplifies the code above a bit.